### PR TITLE
Fix issue with second-level-cache tests and versioned entities

### PR DESCRIPTION
**Pull request by DHager**
I'm not sure why this bug doesn't show normally within `master`, but I ran across it with pull-request build [3930.1](https://travis-ci.org/doctrine/doctrine2/jobs/57975906), where it complained since `$targetPersister` was not an instance of `CachedPersister`. 

The immediate fix seems straightforward, because `DefaultCollectionHydrator` doesn't even accept the second argument.

@FabioBatSilva : Please let me know if you see some better resolution to this, or if it indicates some edge-case in the second-level-cache logic that requires a new test.

### CODE REVIEW

**Comment by FabioBatSilva** [comment]
Thanks for your PR @DHager

If i'm not mistaken,
In order to trigger this error you would have to mark a ``x-to-many`` association as cacheable without marking the target entity as cacheable.

Something like :
```php
/**
 * @Entity
 * @Cache()
 */
class State
{
    /**
     * @Cache
     * @OneToMany(targetEntity="City", mappedBy="state")
     */
    protected $cities;
}

/**
 * @Entity
 */
class City
{
    // ...
}
```

But this is not a valid configuration, the target entity must also be marked as cacheable.

I'm not entirely sure if it will fix the problem because [DefaultCollectionHydrator](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Cache/DefaultCollectionHydrator.php#L78) will try to load the ``CachedPersister`` during the load/hydration process.
In this case we should throw an exception that better explains the miss configuration problem.

Can you add a test for this [see](https://github.com/doctrine/doctrine2/blob/master/tests/Doctrine/Tests/ORM/Functional/SecondLevelCacheOneToManyTest.php#L387)

**Comment by DHager** [comment]
@FabioBatSilva Yeah, that seems to be the case, it's being caused by `CacheMetadataListener`.

The entities being tested initially have no caching at all, but when you add the `ENABLE_SECOND_LEVEL_CACHE` environment var, the listener gets to work turning on caching under certain circumstances. It skips versioned entities, and this works fine as long as all-or-none of them are versioned... but in my #1378 PR, I've enabled versioning just for CompanyEvent and its subclasses.

This leads to a case where `CompanyOrganization` has a caching association, but its destination type `CompanyEvent` is versioned and not cached.

What if `CacheMetadataListener` operated in two passes, first turning on entity-caching, and then doing association-caching only when the association points to entities it hit the first time around?

**Comment by DHager** [comment]
@FabioBatSilva I've rewritten the `CacheMetadataListener` so that it aggressively explores association-types rather than being just a passive listener for activity. Does it still correctly capture the intent of the `ENABLE_SECOND_LEVEL_CACHE` flag?

**Comment by DHager** [comment]
@FabioBatSilva: I'm inclined to call this a issue in the tests themselves, where the "re-do all tests with caching" code was capable of generating an invalid configuration.

I think having better errors when an invalid configuration occurs should be a separate task.

**Comment by goetas** [comment]
This problem has been already solved withhttps://github.com/doctrine/doctrine2/pull/954

**Comment by goetas** [comment]
Wait... @Ocramius the `CacheMetadataListener` changes from #954 has not be merged... Is there a reason?

**Comment by Ocramius** [comment]
@Asmir check the commits that I manually introduced - I pretty much rewrote
the PR during merge operations

Marco Pivetta

http://twitter.com/Ocramius

http://ocramius.github.com/

File: tests/Doctrine/Tests/EventListener/CacheMetadataListener.php [hunk 1]
```diff
@@ -3,33 +3,74 @@
...
+         * Only enable association-caching when the target has already been
+         * given caching settings
+         */
         foreach ($metadata->associationMappings as $mapping) {
-            $metadata->enableAssociationCache($mapping['fieldName'], $cache);
+
+            $targetMeta = $em->getClassMetadata($mapping['targetEntity']);
+            $this->enableCaching($targetMeta, $em);
+
+            if(array_key_exists($targetMeta->getName(), $this->enabledItems)){
```
Comment (goetas) [inline tests/Doctrine/Tests/EventListener/CacheMetadataListener.php]: This check should be redundant since it is already handled at line 46

File: tests/Doctrine/Tests/EventListener/CacheMetadataListener.php [hunk 2]
```diff
@@ -3,33 +3,74 @@
...
+        $this->enableCaching($metadata, $em);
+    }
+
+    /**
+     * @param ClassMetadata $metadata
+     * @param EntityManager $em
+     */
+    protected function enableCaching(ClassMetadata $metadata, EntityManager $em){
+
+        if(array_key_exists($metadata->getName(), $this->enabledItems)){
```
Comment (goetas) [inline tests/Doctrine/Tests/EventListener/CacheMetadataListener.php]: PSR-2,  whitespace after`if`

File: tests/Doctrine/Tests/EventListener/CacheMetadataListener.php [hunk 3]
```diff
@@ -3,33 +3,74 @@
...
-            'usage' => ClassMetadata::CACHE_USAGE_NONSTRICT_READ_WRITE
-        );
+        $em = $event->getObjectManager();
 
         /** @var $metadata \Doctrine\ORM\Mapping\ClassMetadata */
         if (strstr($metadata->name, 'Doctrine\Tests\Models\Cache')) {
             return;
         }
 
+        if( ! $em instanceof EntityManager){
```
Comment (goetas) [inline tests/Doctrine/Tests/EventListener/CacheMetadataListener.php]: is this case possible?

**Comment by goetas** [comment]
@Ocramius yea, you refactored a lot my changes. But, is there a reason for the removal of `CacheMetadataListener` changes. 
I've encountered the same recursion problem highlighted by this PR.

File: tests/Doctrine/Tests/EventListener/CacheMetadataListener.php (hunk 2 above)
Comment (goetas) [inline tests/Doctrine/Tests/EventListener/CacheMetadataListener.php]: Consider `isset` instead of `array_key_exists` (see http://ilia.ws/archives/247-Performance-Analysis-of-isset-vs-array_key_exists.html)

File: tests/Doctrine/Tests/EventListener/CacheMetadataListener.php (hunk 3 above)
Comment (DHager) [inline tests/Doctrine/Tests/EventListener/CacheMetadataListener.php]: AFaIK no current code-path exists, but API-wise nothing's stopping anyone from doing it in the future.

I think we could make it impossible by refactoring the subclass `LoadClassMetadataEventArgs` to use its own constructor with type-hinting. (Not to be confused with its parent class of the same name from a different package.)

**Comment by DHager** [comment]
@goetas : Yeah, it looks like we both ended up with the same idea for changing it to act recursively. 

I made some minor changes to the listener, and added an explicit constructor to `LoadClassMetadataEventArgs` so that type-hinting forces people to only construct it with a real `EntityManager`.

File: lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php [hunk 4]
```diff
@@ -30,12 +32,40 @@
...
+     */
+    function __construct(ClassMetadata $classMetadata, EntityManager $entityManager)
+    {
+        /*
+        We use our own constructor here to enforce type-hinting requirements,
+        since both inputs are specialized subclasses compared to what the super-
+        class is willing to accept.
+
+        In particular, we want to have EntityManager rather than ObjectManager.
+        */
```
Comment (malukenho) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: It's can't be moved to docblock of this method?

Comment (DHager) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: The comment answers the question "Why Even Have This?" that people will ask... but only if they're looking at the seemingly-pointless guts of the method.

From outside the class, nobody needs to know *why* the constructor exists, they just need to know how to use it.

**Comment by DHager** [comment]
It should be fairly low-risk to do this, right? I mean, it's a fix to a testing meta-framework rather than the final product.

If you need a fixture that shows the problem with the test code, I have that in #1378 , which depends on this in order to get a passing build.

**Comment by DHager** [comment]
@Ocramius : Is there more that needs to be done for this fix? Should I cherry-pick over the fixture-changes from #1378 that exhibit the test-framework bug?

File: lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php [hunk 5]
```diff
@@ -30,12 +32,40 @@
...
+    /**
      * Retrieve associated EntityManager.
      *
      * @return \Doctrine\ORM\EntityManager
      */
     public function getEntityManager()
     {
-        return $this->getObjectManager();
+        $em = $this->getObjectManager();
+        assert($em instanceof EntityManager);
```
Comment (guilhermeblanco) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: Please remove this line

Comment (DHager) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: Any particular reason? The function is called  `getEntityManager()` and is explicitly documented as returning an `EntityManager` instance. If that contract ever fails, we probably want to know about it sooner rather than later.

Comment (TomasVotruba) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: @DHager This should be probably covered by unit tests, not runtime asserting.

**Comment by DHager** [comment]
Hello again... so is there anything I can do to finish this thing off, to make it clearer, or to make merging easier? 

Should I cherry-pick over the fixture-changes which reveal the test-bug from #1456, or could that cause merge-conflicts down the road?

File: lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php [hunk 6]
```diff
@@ -30,12 +32,42 @@
 class LoadClassMetadataEventArgs extends BaseLoadClassMetadataEventArgs
 {
     /**
+     * @param ClassMetadata $classMetadata
+     * @param EntityManager $entityManager
+     */
+    function __construct(ClassMetadata $classMetadata, EntityManager $entityManager)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: This is, sadly, a BC break in the API. I'm going to revert this during merge

**Comment by Ocramius** [comment]
Merged, thanks!

`master`: doctrine/doctrine2@34eb4fdec52bf6b897b53342f6df95e1eae1474f
`2.5`: doctrine/doctrine2@12d178777a238b28a725d809e73f6942421e042a

File: lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php (hunk 6 above)
Comment (DHager) [inline lib/Doctrine/ORM/Event/LoadClassMetadataEventArgs.php]: Can we at least deprecate it somehow? Nobody should ever be passing in an `ObjectManager` as the second argument. It wasn't even possible in the initial design, until 8443eee6289314813dae87eb7ca92a3a8811ed31.
